src/main/kotlin/me/partlysanestudios/partlysaneskies/events/render/RenderWaypointEvent.kt

```diff
@@ -0,0 +1,66 @@
+//
+// Written by Su386.
+// See LICENSE for copyright and license notices.
+//
+
+
+package me.partlysanestudios.partlysaneskies.events.render
+
+import me.partlysanestudios.partlysaneskies.events.EventManager
+import me.partlysanestudios.partlysaneskies.render.waypoint.Waypoint
+import net.minecraft.client.Minecraft
+import net.minecraft.util.Vec3
+
+class RenderWaypointEvent(
+    val pipeline: WaypointRenderPipeline
+) {
+
+    companion object {
+        internal fun onEventCall(partialTicks: Float, functions: List<EventManager.EventFunction>) {
+            val pipeline = WaypointRenderPipeline()
+
+            for (function in functions) {
+                val event = RenderWaypointEvent(pipeline)
+                try {
+                    function.function.call(function.obj, event)
+                } catch (exception: Exception) {
+                    exception.printStackTrace()
+                }
+            }
+
+            pipeline.renderAll(partialTicks)
+        }
+    }
+
+    class WaypointRenderPipeline() {
+        private val waypointsToRender = ArrayList<Waypoint>()
+
+        /**
+         * Adds a waypoint to the render pipeline
+         *
+         * @param waypoint the waypoint to be rendered
+         */
+        fun add(waypoint: Waypoint) {
+            waypointsToRender.add(waypoint)
+        }
+
+        // Written partially by j10a1n15
+        /**
+         * Renders all the waypoints in the pipeline
+         *
+         * @param partialTicks the amount of partial ticks since the last frame that rendered
+         */
+        internal fun renderAll(partialTicks: Float) {
+            val mc = Minecraft.getMinecraft()
+            val playerPos = mc.thePlayer.getPositionEyes(partialTicks)
+
+            for (waypoint in waypointsToRender) {
+                val waypointPos =
+                    Vec3(waypoint.position.x.toDouble(), waypoint.position.y.toDouble(), waypoint.position.z.toDouble())
+                val distance = playerPos.distanceTo(waypointPos)
```

Comment:
We could do a full Point3d transition and impl stuff like
Point3d.distanceTo(<other point33>) or just Point3d.distanceToPlayer()

Reply:
We already had this,, it was in math utils. It's been moved

Reply:
I dont think you quite understood,
having a method in Point3d that just returns the distance to player, would make this a bit easier and better future wise

Reply:
yeah but why not .distanceToPlayer() aswell